cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -454,6 +454,114 @@ struct PowerChecked {
   }
 };
 
+// Bitwise operations
+
+struct BitWiseNot {
+  template <typename T, typename Arg>
+  static T Call(KernelContext*, Arg arg, Status*) {
+    return ~arg;
+  }
+};
+
+struct BitWiseAnd {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    return lhs & rhs;
+  }
+};
+
+struct BitWiseOr {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    return lhs | rhs;
+  }
+};
+
+struct BitWiseXor {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    return lhs ^ rhs;
+  }
+};
+
+struct ShiftLeft {
+  template <typename T, typename Arg0, typename Arg1>
+  static T Call(KernelContext*, Arg0 lhs, Arg1 rhs, Status*) {
+    using Unsigned = typename std::make_unsigned<Arg0>::type;
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE(rhs < 0 || rhs >= std::numeric_limits<Arg0>::digits)) {
+      return lhs;
+    }
+    return static_cast<T>(static_cast<Unsigned>(lhs) << static_cast<Unsigned>(rhs));
+  }
+};
+
+// See SEI CERT C Coding Standard rule INT34-C
+struct ShiftLeftChecked {
+  template <typename T, typename Arg0, typename Arg1>
+  static enable_if_unsigned_integer<T> Call(KernelContext*, Arg0 lhs, Arg1 rhs,
+                                            Status* st) {
+    static_assert(std::is_same<T, Arg0>::value, "");
+    if (ARROW_PREDICT_FALSE(rhs < 0)) {
+      *st = Status::Invalid("shift must be non-negative");
+      return lhs;
+    }
+    if (ARROW_PREDICT_FALSE(rhs >= std::numeric_limits<Arg0>::digits)) {
+      *st = Status::Invalid("overflow");
+      return lhs;
+    }
+    return lhs << rhs;
```

Comment:
It's a bit weird not to check for overflow here. Is it easily doable? (perhaps by checking that `(lhs << rhs) >> rhs == lhs`?)

Reply:
I thought for shifts, usually it's desired to take the remainder if there would be overflow? Though [Wikipedia actually differentiates between arithmetic and logical left shift](https://en.wikipedia.org/wiki/Arithmetic_shift#Equivalence_of_arithmetic_and_logical_left_shifts_and_multiplication) here: the former raises overflow errors, the latter returns the modulo.

(Also, I think it's just sufficient to check that (lhs << rhs) < lhs.)

Reply:
In which context do people want to do left or right shifts exactly? That doesn't sound like a common data analysis task. @ianmcook

Reply:
> (Also, I think it's just sufficient to check that (lhs << rhs) < lhs.)

I don't think so, for example:
```python
>>> (36 << 4) & 255
64
```

Reply:
D'oh, fair enough.

Looking at Postgres, the example they give just wraps around, at least: https://www.postgresql.org/docs/current/functions-bitstring.html
MSSQL/T-SQL don't support shifts, even.

Reply:
Hmm, the Postgres results are... interesting.
If you enter this request [here](https://extendsclass.com/postgresql-online.html):
```sql
drop table if exists test;
create table test (a int2);
insert into test values (9263);
select (a << 4) a4, (a << 12) a12, (a << 123) a123, (a << -1) am1 from test;
```
you get the tuple `17136 | -4096 | 0 | 0`.

Reply:
The first two results seem to make sense. The last two would be undefined so I'm not surprised you get odd results (that said, I can see why you might expect a very large shift to result in 0).

Reply:
Amusingly, MySQL does seem to check for overflows.
```sql
drop table if exists shifted;
drop table if exists test;
create table test (a smallint);
insert into test values (9263);
create table shifted (b smallint);
insert into shifted select (a << 3) from test;
select * from shifted;
```
[gives](https://extendsclass.com/mysql-online.html):
```
Out of range value for column 'b' at row 1
```